[review-round-2 - Jan 3 merge] What is a Python package intro tutorial? (#139)
pyOpenSci/python-package-guide: tutorials/intro.md

content will still be valuable if you are interested in better
understanding the steps involved in creating a Python package.

* In this series you will learn about the core elements that you need to publish your package to the [Python Package Index (PyPI)](https://pypi.org/).

* In the second series, you will learn about infrastructure and documentation needed to support package maintenance.

:::{toctree}
:hidden:
:caption: Python Packaging 101

What is a Python package? <self>
Make your code pip installable <1-installable-code>
Add README file <2-add-readme>
Add LICENSE & CODE of CONDUCT <4-add-license-file>
Add pyproject.toml <5-pyproject-toml>
Publish to PyPI <6-publish-pypi>
Publish to conda-forge <7-publish-conda-forge>
:::

:::{admonition} Learning Objectives

This lesson introduces you to the basic components of a Python package.
After reading this lesson you will:

- Understand what a Python package is
- Be able to list the 5 core components of a Python package
- Be able to explain the difference between generalizable code and code that supports a specific scientific application

:::

:::{todo}
You might also hear people use the term "library" the same way.

> :::

Reviewer: looks like `{todo}` doesn't get rendered, so I'm guessing it's a note -- is this supposed to be addressed in this PR?

Author: good catch. i removed this!

## What is a Python package?

At a high level, you can think about a Python package as a toolbox
that you can use to perform various tasks.

A Python package is basically a directory with a specific
file structure. Within the package directory structure, there are
modules which are files that end in `.py` (the same extension you'd
see in a Python script). These modules allow you to group and
structure your Python code. Each module contains functions and classes,
that you can think about as the tools in your toolbox.

:::{figure-md} python-toolbox

<img src="../images/tutorials/toolbox.png" alt="Diagram showing a sketch of a toolbox filled with different tools including a hammer and a saw." width="400px">

You can think about a package as a toolbox filled with coding tools.
A tool may be a function or a class. Each tool does a specific thing
well.
:::

### Python packages are installable

A package is installable, which means that you can add the
functionality within the package's code to any Python environment and

> import that functionality like you would import core Python packages such

Reviewer: ```suggestion
import that functionality like you would import core scientific Python packages such
```
nit ("core python" feels confusable with "python standard library", especially for beginners)

Author: oh you're right. good catch!

as NumPy or Matplotlib.

```python
import numpy
```

Installing a package into an environment makes it easier
to manage and reuse your code across different projects.
Structuring your code as a package is the first step you need to take so you can share the tools in the toolbox you've created and let others build with it.

## Why create a Python package?

You might create a Python package because you want to:

- **Use your code across different projects:** At its most basic level, creating a package allows you to install your code into a Python environment. This allows you to then import functions and classes into any workflows both locally and in the cloud.
- **Share your code:** If you publish a package on a public repository such as PyPI or conda, your package can be installed on any machine using pip or conda with a single command.

> - **Build community around your code:** Packages make it easier for multiple people to work on the same project (particularly when published on GitHub). A version platform such as git (the version control system used by GitHub), further makes it easier to track changes to the codebase over time. Tools such as issues and pull requests make it easier for outside users to contribute bug fixes and to establish review processes for accepting changes to the code base.

Reviewer: ```suggestion
- **Build community around your code:** Packages make it easier for multiple people to work on the same project (particularly when published on GitHub). A version control system such as git (the system used by GitHub), further makes it easier to track changes to the codebase over time. Tools such as issues and pull requests make it easier for outside users to contribute bug fixes and to establish review processes for accepting changes to the code base.
```
nit (git's not a platform, right?)

> - **Build community around your code:** Packages make it easier for multiple people to work on the same project (particularly when published on GitHub). A version platform such as git (the version control system used by GitHub), further makes it easier to track changes to the codebase over time. Tools such as issues and pull requests make it easier for outside users to contribute bug fixes and to establish review processes for accepting changes to the code base..

Author: They aren't intentional so thank you for catching them! i'm not sure how i did that so many times

- **Organize your code:** Packages can be used to organize large code projects, dividing them into smaller, more manageable components. This structure can help with both maintaining the codebase and with making it easier to understand.

### What to consider before you create a package

Creating a Python package that others use takes considerable
time and effort. Before you begin, think about your goals including:

- Who you think will use your package
- How people might use your package and on what data (if data are relevant)
- Whether you have time to add things such as documentation and tests
- How long you might be able to maintain it: remember that once people begin using your package they will depend on your maintainer team to update it, fix bugs and answer questions.

Before creating a user-facing package, it's important to consider all of the above.

## The elements of a Python package

:::{figure-md} package-components

<img src="../images/tutorials/package-components.png" alt="Diagram showing .. more here if this stays." width="500px">

The elements of a Python package include code, documentation, tests,

> an OSI-approved license and a specific directory structure.

Reviewer: the image says "infrastructure", while the caption says "directory structure". they should be the same, right?

Author: excellent catch! i updated the graphic but forgot to update the caption!! they should be the same! this is fixed locally.

Maintainers are at the core making sure everything works and is up to
date while fixing bugs and addressing user concerns.
:::

The core elements of Python package include:

- **Code:** Functions and classes that provide functionality for a user of your package
- **Documentation:** Installation instructions, tutorials, and examples that both help users get started using your package and contributors and maintainers fix bugs and maintain the package.
  - Contributor Documentation in the form of a **CONTRIBUTING.md** file is useful to help people to contribute to your package.

>   - Development documentation helps both maintainers and contributors understand how to maintain a package's infrastructure.

Reviewer 2: ```suggestion
  - Development Documentation helps both maintainers and contributors understand how to maintain a package's infrastructure.
```

> - **Tests:** that makes sure your code works as it should and makes it easier for you and others to contribute to, modify and update the code in the future

Reviewer 2: ```suggestion
- **Tests:** that make sure your code works as it should and makes it easier for you and others to contribute to, modify and update the code in the future
```

> - **License:** An open source license, or license that is [OSI approved](https://opensource.org/licenses/) refers to an license that allows others to use your package. It also provides legal direction regarding how elements of the package can and can't be reused.

Reviewer: ```suggestion
- **License:** An open source license, or license that is [OSI approved](https://opensource.org/licenses/), refers to an license that allows others to use your package. It also provides legal direction regarding how elements of the package can and can't be reused.
```

> - **License:** An open source license, or license that is [OSI approved](https://opensource.org/licenses/) refers to an license that allows others to both use your package. It also provides legal direction regarding how elements of the package can and can't be reused.

Reviewer 2: ```suggestion
- **License:** An open source license, or license that is [OSI approved](https://opensource.org/licenses/) refers to an license that allows others to use your package. It also provides legal direction regarding how elements of the package can and can't be reused.
```

- **Infrastructure** that automates updates, publication workflows and runs test suites. Infrastructure includes a suite of things such as platforms like GitHub and GitLab, tools to run tests and tools locally such as nox and tox and continuous integration that automates package maintenance steps.

:::{admonition} What pyOpenSci looks for in a package
:class: tip

pyOpenSci performs an [initial set of editor checks](https://www.pyopensci.org/software-peer-review/how-to/editor-in-chief-guide.html#editor-checklist-template) for any package submitted
to us for peer review.
You may find these checks useful as you create your package as a baseline for
things that you package should have.
:::

## Packages are more than just code - Infrastructure

A package in any language is more than just code. If you expect other
people to use your package, besides yourself, you should consider not
only writing high quality code, but also the various elements of a
package that make it a useful community resource.

### Version control and storing your package on GitHub or GitLab

Most Python packages live in an online version control platform such as GitHub

> or GitLab. GitHub and GitLab are both

Reviewer 2: ```suggestion
or GitLab. GitHub and GitLab both
```

under version control is important because it allows you to both track changes over time while also going back in history and undoing changes in the case that a change to the code base unexpectedly breaks something.

By publishing your package on GitHub or GitLab, you are making your code public facing. This means that others can both see your code and also make contributions using a pull request (GitHub) / merge request (GitLab) /  code review workflow.

:::{admonition} GitHub & GitLab vs. Git
:class: tip

GitHub and GitLab are online (cloud) platforms that run `git` (version control
software) on the backend. Running git locally on your computer allows you to
upload (`git push`) and download (`git pull`) files to GitHub and GitLab.
:::

### Issues or Ticket Trackers

GitHub and GitLab also both offer community features such as issues that allow:

* you to communicate with your maintainers and contributor community
* users to report bugs, ask questions and request new features
* you to publicly keep track of enhancements and features you want to work on for your package.

### Continuous integration and continuous deployment

> GitHub and GitLab also provide continuous integration and continuous deployment (CI/CD). Continuous integration (CI) refers to a platform that automatically runs a specific job when a certain event occurs.

Reviewer: ```suggestion
GitHub and GitLab also provide continuous integration and continuous deployment (CI/CD). Continuous integration (CI) refers to a platform that automatically runs a specific job when a certain event occurs, whereas continuous deployment (CD) is a form of CI that runs before publication.
```
or a better description from someone who understands CD, but there should be a parallel description of CD here.

Author: ahhh ok. Bill i'm going to modify this suggest just a bit. it runs in order to deploy / publish - ie publish to pypi or publish to readthedocs or publish to github pages . i'll commit this and then will make a minor fix but it's a very good suggestion!!

> GitHub and GitLab also provide continuous integration and continuous deployment (CI/CD). Continuous integration (CI) refers to a platform that automatically runs a specific job when a certain event occurs, whereas continuous deployment (CD) is a form of CI that runs before publication.

Author: ```suggestion
GitHub and GitLab also provide continuous integration and continuous deployment (CI/CD). Continuous integration (CI) refers to a platform that automatically runs a specific job when a certain event occurs, whereas continuous deployment (CD) is an extension of CI that refers to not only running or building but also to publishing the final outputs somewhere. 
```

change to your code.

:::{figure-md} packaging-workflow

<img src="../images/tutorials/packaging-lifecycle.png" alt="Diagram showing .. more here if this stays." width="700px">

The lifecycle of a scientific Python package.
:::

## When should you turn your code into a Python package?

You may be wondering, what types of code should become a Python package that is both on GitHub and published to PyPI and/or conda-forge.

There are a few use cases to consider:

1. **Creating a basic package for yourself:** Sometimes you want create a package for your own personal use. This might mean making your code locally pip installable and you may also want to publish it to GitHub. In that case you don't expect others to use your code, and as such you may only have documentation for you and your future self if you need to update the package.

> An example of this type of package might be a set of functions that you write that are useful across several of your projects. It could be useful to have those functions available to all of your projects.

:::{todo}
LINK to pip installable lesson when it's published - it's in review now
:::

> 2. In other cases, you may create some code that you soon realize might also be useful to not just you, but to other people as well.

Reviewer 2: ```suggestion
2. **Creating a package for the community:** In other cases, you may create some code that you soon realize might also be useful to not just you, but to other people as well.
```

> In that case, you might consider both creating the package, publishing it on GitHub, and because other users may be using it, you may make user of GitHub's infrastructure including CI/CD pipelines, issue trackers. Because you want other people to use your package, you will want to also include LICENSE information, documentation for users and contributors and tests. This type of package is most often published to PyPI.

Reviewer 2: ```suggestion
In that case, you might consider both creating the package, publishing it on GitHub, and because other users may be using it, you may make use of GitHub's infrastructure including CI/CD pipelines and issue trackers. Because you want other people to use your package, you will want to also include LICENSE information, documentation for users and contributors and tests. This type of package is most often published to PyPI.
```

Matplotlib does one thing really well:

*It creates visual plots of data.*

Thousands of people use Matplotlib for different plotting applications
using different types of data. While few scientific packages will have the same
broad application and large user base that Matplotlib has, the
idea of scoping out what your package does is still important.
:::

### Code should also be clean & readable & documented

The code in your package should also be clean, readable, and well documented.

**Clean code:** Clean code refers to code that uses expressive variable names,
is concise and doesn't repeat itself. You can learn about best practices
for clean code in future pyOpenSci tutorials.

**Readable code:** readable code is code written with a consistent style.
You can use linters and code formatters such as black and flake8 to ensure
this consistency throughout your entire package. [Learn more about code formatters here.](../package-structure-code/code-style-linting-format.html)

**Documented code:** documented code is written using docstrings that help a
user understand both what the functions and methods in your code do and also

> what the input and output elements of each function is. [You can learn more about docstrings in our guide, here.](../documentation/write-user-documentation/document-your-code-api-docstrings)

Reviewer 2: ```suggestion
what the input and output elements of each function are. [You can learn more about docstrings in our guide, here.](../documentation/write-user-documentation/document-your-code-api-docstrings)
```

* [a development guide that documents your maintainer workflow process ](/documentation/repository-files/development-guide.md)
* [a code of conduct to defines community interaction standards and expectations](/documentation/repository-files/code-of-conduct-file.md)
* [a contributing guide that helps users understand expectations associated with making contributions to your project](/documentation/repository-files/contributing-file.md)

:::{admonition} Support for contributors and maintainers

If you intend for others to use and contribute to your code, consider who will maintain it over time. You will want a **contributing and development** guide to help new potential contributors get started with contributing to your package, as well as a **code of conduct** to ensure community interactions remain healthy both for you and your contributors and maintainer team.

The elements above are also important for future maintenance of your package. In the case that you are no long able to maintain it or simply want extra help, development, and contributing documentation will help you onboard new maintainers.
:::

*****

## What's next?

> In future lessons you will learn talk more about the infrastructure around a published Python package that makes it both easier to maintain, easier for others to contribute to and easier for other scientists to use. However, first we want to get you to your initial goal of publishing a Python package.

Reviewer 2: ```suggestion
In future lessons you will learn more about the infrastructure around a published Python package that makes it both easier to maintain, easier for others to contribute to and easier for other scientists to use. However, first we want to get you to your initial goal of publishing a Python package.
```
